src/drivers/fpga/xrt/selftests/test1-main.c

```diff
@@ -0,0 +1,281 @@
+		xleaf_put_leaf(xdd->pdev, xdd->leaf0);
+		return xdd;
+	}
+
+	xrt_info(pdev, "xleaf test instance[0] %p", xdd->leaf0);
+	xrt_info(pdev, "xleaf test instance[1] %p", xdd->leaf1);
+	return xdd;
+}
+
+/* Basic test for XRT core which validates inter xleaf ioctl calls. Symbolically we
+ * are performing the following operation:
+ *
+ * group2.xmgmt_main() {
+ *     ioctl(group0.test, XRT_XLEAF_TEST_A, arg)
+ *     ioctl(group1.test, XRT_XLEAF_TEST_B, arg) {
+ *         lookup(group0.test)
+ *         ioctl(group0.test, XRT_XLEAF_TEST_A, arg)
+ *     }
+ * }
+ */
+static int test1_validate_fini(struct test1_main_client_data *xdd)
+{
+	int ret;
+	struct xrt_xleaf_test_payload arg_a = {uuid_null, "FPGA"};
+	struct xrt_xleaf_test_payload arg_b = {uuid_null, "FPGA"};
+
+	if (!xdd->leaf1)
+		goto finally;
+
+	generate_random_uuid(arg_a.dummy1.b);
+	generate_random_uuid(arg_b.dummy1.b);
+	ret = xleaf_ioctl(xdd->leaf0, XRT_XLEAF_TEST_A, &arg_a);
+	if (ret || !uuid_is_null(&arg_a.dummy1) || strcmp(arg_a.dummy2, "alveo")) {
+		xrt_err(xdd->pdev, "xleaf test instance[0] %p ioctl %d failed", xdd->leaf1, XRT_XLEAF_TEST_A);
+		ret = -EDOM;
+		goto error;
+	}
+	ret = xleaf_ioctl(xdd->leaf1, XRT_XLEAF_TEST_B, &arg_b);
+	if (ret || !uuid_is_null(&arg_b.dummy1) || strcmp(arg_b.dummy2, "alveo")) {
+		xrt_err(xdd->pdev, "xleaf test instance[1] %p ioctl %d failed", xdd->leaf1, XRT_XLEAF_TEST_B);
+		ret = -EDOM;
+		goto error;
+	}
+
+error:
+	xleaf_put_leaf(xdd->pdev, xdd->leaf1);
+	xleaf_put_leaf(xdd->pdev, xdd->leaf0);
+finally:
+	vfree(xdd);
+	return ret;
+}
+
+static int test1_main_leaf_ioctl(struct platform_device *pdev, u32 cmd, void *arg)
+{
+	struct test1_main *xmm = platform_get_drvdata(pdev);
+	int ret = 0;
+
+	xrt_info(pdev, "%p.ioctl(%d, %p)", xmm, cmd, arg);
+	switch (cmd) {
+	case XRT_XLEAF_EVENT:
+		test1_main_event_cb(pdev, arg);
+		break;
+	default:
+		xrt_err(pdev, "unknown cmd: %d", cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static ssize_t test1_main_leaf_read(struct file *file, char __user *ubuf,
+				    size_t n, loff_t *off)
+{
+	int i;
+	struct test1_main_client_data *xdd = file->private_data;
+
+	for (i = 0; i < 4; i++) {
+		xrt_info(xdd->pdev, "reading...");
+		ssleep(1);
+	}
+	return n;
+}
+
+static ssize_t test1_main_leaf_write(struct file *file, const char __user *ubuf,
+				     size_t n, loff_t *off)
+{
+	int i;
+	struct test1_main_client_data *xdd = file->private_data;
+
+	for (i = 0; i < 4; i++) {
+		xrt_info(xdd->pdev, "writing %d...", i);
+		ssleep(1);
+	}
+	return n;
+}
+
+static int test1_main_open(struct inode *inode, struct file *file)
+{
+	struct platform_device *pdev = xleaf_devnode_open(inode);
+	struct test1_main_client_data *xdd;
+
+	/* Device may have gone already when we get here. */
+	if (!pdev)
+		return -ENODEV;
+
+	xrt_info(pdev, "opened");
+	/* Obtain the reference to test xleaf nodes */
+	xdd = test1_validate_ini(pdev);
+	file->private_data = xdd;
+	return xdd->leaf1 ? 0 : -EDOM;
```

Comment:
do you want to vfree(xdd) when you return -EDOM?

Reply:
xdd saves pdev for use later in fini, hence not freeing it here.

Reply:
if you return -EDOM, open will fail and you'd expect a corresponding close will still happen and give you a chance to call _fini to clean up?

Reply:
Good catch, I had updated the code to return error code to user space so fini will not run in error path. Let me fix this.

Reply:
Cleanup should happen now.